File: cpp/173_BinarySearchTreeIterator.cpp
/**
 * Definition for binary tree
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class BSTIterator {
public:
    BSTIterator(TreeNode *root) {
        for(auto p = root; p != NULL; p = p->left){
            sTree.push(p);
        }
    }

    /** @return whether we have a next smallest number */
    bool hasNext() {
        return !(sTree.empty());
    }

    /** @return the next smallest number */
    int next() {
        auto tmp = sTree.top();
        sTree.pop();
        for(auto q = tmp->right; q != NULL; q = q->left){
            sTree.push(q);
        }
        return tmp->val;
    }
private:
    stack<TreeNode*> sTree;
};

/**
 * Your BSTIterator will be called like this:
 * BSTIterator i = BSTIterator(root);
 * while (i.hasNext()) cout << i.next();
 */

File: js/173_BinarySearchTreeIterator.js
/**
 * Definition for binary tree
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */

/**
 * @constructor
 * @param {TreeNode} root - root of the binary search tree
 */
var BSTIterator = function(root) {
    this.sTree = [];
    for(var p = root; p !== null; p = p.left){
        this.sTree.push(p);
    }
    console.log(this.sTree);
};

/**
 * @this BSTIterator
 * @returns {boolean} - whether we have a next smallest number
 */
BSTIterator.prototype.hasNext = function() {
    return this.sTree.length > 0;
};

/**
 * @this BSTIterator
 * @returns {number} - the next smallest number
 */
BSTIterator.prototype.next = function() {
    var tmp = this.sTree.pop();
    for(var q = tmp.right; q !== null; q = q.left){
        this.sTree.push(q);
    }
    return tmp.val;
};

/**
 * Your BSTIterator will be called like this:
 * var i = new BSTIterator(root), a = [];
 * while (i.hasNext()) a.push(i.next());
*/